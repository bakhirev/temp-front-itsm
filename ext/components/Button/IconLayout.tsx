import styled, { css } from 'styled-components';

const ICON_LAYOUT_MARGIN_HORIZONTAL = 6;

interface IIconLayoutProps {
  onlyIcon: boolean;
  rightIcon: boolean;
}

export const IconLayout = styled.div<IIconLayoutProps>`
  display: flex;
  ${({ onlyIcon, rightIcon }) =>
    !onlyIcon
      ? css`
          margin-left: ${rightIcon ? `${ICON_LAYOUT_MARGIN_HORIZONTAL}px` : 0};
          margin-right: ${!rightIcon ? `${ICON_LAYOUT_MARGIN_HORIZONTAL}px` : 0};
        `
      : ''};
`;
