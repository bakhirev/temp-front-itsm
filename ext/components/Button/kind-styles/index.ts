import { primary } from './primary';
import { secondary } from './secondary';
import { ghost } from './ghost';
import { white } from './white';
import { danger } from './danger';
import { success } from './success';

export const KIND_STYLES = {
  primary,
  secondary,
  ghost,
  white,
  danger,
  success,
};
