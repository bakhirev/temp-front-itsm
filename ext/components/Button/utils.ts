import type { ITheme } from '../themes';

import type { Size } from './ButtonComponent';

const BORDER_WIDTH = 1;
const PADDING_COMPENSATION = 2;

export const BUTTON_SIZE = {
  big: 56,
  medium: 48,
  small: 40,
  micro: 32,
};

export const calculatePadding = ({
  multipleChildren,
  rightIcon,
  size,
  theme: {
    button: { paddingHorizontal },
  },
}: {
  multipleChildren: boolean;
  rightIcon: boolean;
  size: Size;
  theme: ITheme;
}) => {
  const padding = paddingHorizontal[size] - BORDER_WIDTH;

  if (!multipleChildren) return `0 ${padding}px`;

  const left = padding - (rightIcon ? 0 : PADDING_COMPENSATION);
  const right = padding - (rightIcon ? PADDING_COMPENSATION : 0);

  return `0 ${right}px 0 ${left}px`;
};

export const getWidth = ({
  fullWidth,
  onlyIcon,
  size,
}: {
  fullWidth: boolean;
  onlyIcon: boolean;
  size: Size;
}) => {
  if (onlyIcon) return `${BUTTON_SIZE[size]}px`;
  if (fullWidth) return '100%';
  return 'auto';
};
