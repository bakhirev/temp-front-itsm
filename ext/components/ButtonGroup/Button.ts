import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { getKindStyle } from './getKindStyle';
import type { Size } from './ButtonGroupComponent';

const PADDING_HORIZONTAL = {
  medium: 24,
  small: 20,
  micro: 16,
};

const BORDER_WIDTH = 1;

export type Kind = 'solid' | 'outlined';

interface IButtonProps {
  active: boolean;
  disabled: boolean;
  kind: Kind;
  size: Size;
}

export const Button = styled.div<IButtonProps>`
  ${({ active, disabled, kind, size, theme }) => css`
    align-items: center;
    border-radius: ${theme.borderRadius};
    border-style: solid;
    border-width: ${BORDER_WIDTH}px;
    box-sizing: border-box;
    color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    cursor: ${disabled ? 'default' : 'pointer'};
    display: flex;
    padding: 0 ${PADDING_HORIZONTAL[size] - BORDER_WIDTH}px;
    position: relative;

    & div svg {
      fill: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    }

    &:focus {
      outline: none;

      &::after {
        border-radius: ${theme.borderRadius};
        border-style: solid;
        border-width: 1px;
        bottom: 2px;
        content: '';
        display: block;
        left: 2px;
        position: absolute;
        right: 2px;
        top: 2px;
      }
    }

    ${getKindStyle({ active, disabled, kind, theme })};
  `};
`;

Button.defaultProps = {
  theme: DEFAULT_THEME,
};
