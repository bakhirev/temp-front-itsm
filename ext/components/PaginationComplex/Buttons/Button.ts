import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../common';

const SIZE = 40;

export const Button = styled.button`
  ${({ theme }) => css`
    align-items: center;
    background: ${theme.color.neutral[10]};
    border-radius: ${theme.borderRadius};
    border: none;
    cursor: pointer;
    display: flex;
    height: ${SIZE}px;
    justify-content: center;
    outline: none;
    width: ${SIZE}px;

    & svg {
      fill: ${theme.color.neutral[50]};
    }

    &:focus {
      border: 2px solid ${theme.color.primary[60]};
    }

    &:hover {
      background: ${theme.color.neutral[20]};

      & svg {
        fill: ${theme.color.primary[60]};
      }
    }

    &:active {
      background: ${theme.color.neutral[30]};

      & div svg {
        fill: ${theme.color.primary[60]};
      }
    }

    &:disabled {
      background: ${theme.color.neutral[10]};
      cursor: default;

      & svg {
        fill: ${theme.color.neutral[30]};
      }
    }
  `};
`;

Button.defaultProps = {
  theme: DEFAULT_THEME,
};
