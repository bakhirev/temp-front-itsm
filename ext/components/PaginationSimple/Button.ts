import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

const SIZE = 40;

interface IButtonProps {
  selected?: boolean;
  disabled?: boolean;
  withEllipsis?: boolean;
}

export const Button = styled.button<IButtonProps>`
  ${({ disabled, selected, theme, withEllipsis }) => css`
    align-items: center;
    background: ${selected ? theme.color.primary[60] : 'transparent'};
    border-radius: ${theme.borderRadius};
    border: none;
    color: ${selected ? theme.color.neutral.white : theme.color.neutral[90]};
    cursor: ${withEllipsis ? 'default' : 'pointer'};
    display: flex;
    height: ${SIZE}px;
    justify-content: center;
    outline: none;
    user-select: none;
    width: ${SIZE}px;

    & svg {
      fill: ${theme.color.neutral[90]};
    }

    &:focus {
      background: ${disabled || selected || withEllipsis ? '' : theme.color.opacity.blackFocus};
    }

    &:hover {
      background: ${disabled || selected || withEllipsis ? '' : theme.color.opacity.blackHover};
    }

    &:active {
      background: ${disabled || selected || withEllipsis ? '' : theme.color.opacity.blackPressed};
    }

    &:disabled {
      background: transparent;
      color: ${theme.color.neutral[30]};
      cursor: default;

      & svg {
        fill: ${theme.color.neutral[30]};
      }
    }
  `};
`;

Button.defaultProps = {
  theme: DEFAULT_THEME,
};
