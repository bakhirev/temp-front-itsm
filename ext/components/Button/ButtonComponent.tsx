import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { KIND_STYLES } from './kind-styles';
import { calculatePadding, getWidth, BUTTON_SIZE } from './utils';

export type Kind = 'primary' | 'secondary' | 'ghost' | 'white' | 'danger' | 'success';
export type Size = 'big' | 'medium' | 'small' | 'micro';
export type ButtonBasicProps = React.ButtonHTMLAttributes<HTMLButtonElement>;

interface IButtonComponentProps extends ButtonBasicProps {
  fullWidth: boolean;
  kind: Kind;
  multipleChildren: boolean;
  onlyIcon: boolean;
  rightIcon: boolean;
  size: Size;
}

export const ButtonComponent = styled.button<IButtonComponentProps>`
  ${({ fullWidth, kind, multipleChildren, onlyIcon, rightIcon, size, theme }) => css`
    align-items: center;
    border-radius: ${theme.borderRadius};
    border-style: solid;
    border-width: 1px;
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    height: ${BUTTON_SIZE[size]}px;
    justify-content: center;
    outline: none;
    padding: ${onlyIcon ? 0 : calculatePadding({ multipleChildren, rightIcon, size, theme })};
    position: relative;
    width: ${getWidth({ fullWidth, onlyIcon, size })};
    ${KIND_STYLES[kind](theme)};

    &:focus {
      &::after {
        border-radius: ${theme.borderRadius};
        border-style: solid;
        border-width: 1px;
        bottom: 2px;
        content: '';
        display: block;
        left: 2px;
        position: absolute;
        right: 2px;
        top: 2px;
      }
    }

    &:disabled {
      cursor: default;
    }
  `};
`;

ButtonComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
