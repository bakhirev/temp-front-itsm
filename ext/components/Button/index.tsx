import React from 'react';
import type { FC, MouseEvent, ReactNode } from 'react';

import { Button1, Button2 } from '../Typography';

import { IconLayout } from './IconLayout';
import { ButtonComponent } from './ButtonComponent';
import type { ButtonBasicProps, Kind, Size } from './ButtonComponent';

type Type = 'button' | 'submit' | 'reset';

export interface IButtonProps extends ButtonBasicProps {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Отключение кнопки */
  disabled?: boolean;
  /** Растягивает кнопку на всю ширину контейнера */
  fullWidth?: boolean;
  /** Внешний вид кнопки */
  kind?: Kind;
  /** Размер кнопки */
  size?: Size;
  /** Тип кнопки */
  type?: Type;
  /** Коллбэк на клик */
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
}

export const Button: FC<IButtonProps> = ({
  children,
  className,
  dataTestId,
  disabled = false,
  fullWidth = false,
  kind = 'primary',
  size = 'medium',
  type = 'button',
  onClick,
  ...props
}) => {
  const isObject = (element: ReactNode) => typeof element === 'object';

  const onlyIcon = Boolean(children) && !Array.isArray(children) && isObject(children);
  const rightIcon = Array.isArray(children) && isObject(children[children.length - 1]);

  const Text = size === 'big' || size === 'medium' ? Button1 : Button2;

  const handleFocusPrevent = (event: MouseEvent<HTMLButtonElement>) => event.preventDefault();

  const renderChild = (child: ReactNode, key?: number) =>
    isObject(child) ? (
      <IconLayout key={key} onlyIcon={onlyIcon} rightIcon={rightIcon}>
        {child}
      </IconLayout>
    ) : (
      <Text key={key}>{child}</Text>
    );

  const renderChildren = () =>
    Array.isArray(children) ? children.map(renderChild) : renderChild(children);

  return (
    <ButtonComponent
      {...props}
      className={className}
      data-test-id={dataTestId}
      disabled={disabled}
      fullWidth={fullWidth}
      kind={kind}
      multipleChildren={Array.isArray(children)}
      onlyIcon={onlyIcon}
      rightIcon={rightIcon}
      size={size}
      type={type}
      onClick={onClick}
      onMouseDown={handleFocusPrevent}
    >
      {children && renderChildren()}
    </ButtonComponent>
  );
};
