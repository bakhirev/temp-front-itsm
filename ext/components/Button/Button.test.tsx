import React from 'react';
import { shallow } from 'enzyme';

import { Button } from '../Button';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('Button', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const Icon = () => <div />;

  it('should render component', () => {
    const wrapper = shallow(<Button />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with text', () => {
    const wrapper = shallow(<Button>Text</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with icon', () => {
    const wrapper = shallow(
      <Button>
        <Icon />
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with text and left icon', () => {
    const wrapper = shallow(
      <Button>
        <Icon /> Text
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with text and right icon', () => {
    const wrapper = shallow(
      <Button>
        Text <Icon />
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onClick when user clicks on component', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick} />);
    wrapper.simulate('click');
    expect(onClick).toBeCalledTimes(1);
  });
});
