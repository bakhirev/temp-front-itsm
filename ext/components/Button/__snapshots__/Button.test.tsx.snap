// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render component 1`] = `
<styled.button
  disabled={false}
  fullWidth={false}
  kind="primary"
  multipleChildren={false}
  onMouseDown={[Function]}
  onlyIcon={false}
  rightIcon={false}
  size="medium"
  theme={Object {}}
  type="button"
/>
`;

exports[`Button should render component with icon 1`] = `
<styled.button
  disabled={false}
  fullWidth={false}
  kind="primary"
  multipleChildren={false}
  onMouseDown={[Function]}
  onlyIcon={true}
  rightIcon={false}
  size="medium"
  theme={Object {}}
  type="button"
>
  <styled.div
    onlyIcon={true}
    rightIcon={false}
  >
    <Icon />
  </styled.div>
</styled.button>
`;

exports[`Button should render component with text 1`] = `
<styled.button
  disabled={false}
  fullWidth={false}
  kind="primary"
  multipleChildren={false}
  onMouseDown={[Function]}
  onlyIcon={false}
  rightIcon={false}
  size="medium"
  theme={Object {}}
  type="button"
>
  <styled.div>
    Text
  </styled.div>
</styled.button>
`;

exports[`Button should render component with text and left icon 1`] = `
<styled.button
  disabled={false}
  fullWidth={false}
  kind="primary"
  multipleChildren={true}
  onMouseDown={[Function]}
  onlyIcon={false}
  rightIcon={false}
  size="medium"
  theme={Object {}}
  type="button"
>
  <styled.div
    key="0"
    onlyIcon={false}
    rightIcon={false}
  >
    <Icon />
  </styled.div>
  <styled.div
    key="1"
  >
     Text
  </styled.div>
</styled.button>
`;

exports[`Button should render component with text and right icon 1`] = `
<styled.button
  disabled={false}
  fullWidth={false}
  kind="primary"
  multipleChildren={true}
  onMouseDown={[Function]}
  onlyIcon={false}
  rightIcon={true}
  size="medium"
  theme={Object {}}
  type="button"
>
  <styled.div
    key="0"
  >
    Text 
  </styled.div>
  <styled.div
    key="1"
    onlyIcon={false}
    rightIcon={true}
  >
    <Icon />
  </styled.div>
</styled.button>
`;
